base/array.jl

```diff
@@ -2162,35 +2162,35 @@ indmin(a) = findmin(a)[2]
 """
     indexin(a, b)
 
-Return a vector containing the highest index in `b` for
-each value in `a` that is a member of `b` . The output
-vector contains 0 wherever `a` is not a member of `b`.
+Return an array containing the highest index in `b` for
+each value in `a` that is a member of `b`. The output
+array contains `nothing` wherever `a` is not a member of `b`.
 
 # Examples
 ```jldoctest
-julia> a = ['a', 'b', 'c', 'b', 'd', 'a'];
+julia> a = ['a', 'b', 'c', 'b', 'd', 'a']
 
-julia> b = ['a','b','c'];
+julia> b = ['a', 'b', 'c']
 
-julia> indexin(a,b)
-6-element Array{Int64,1}:
+julia> indexin(a, b)
+6-element Array{Union{Nothing, Int64},1}:
  1
  2
  3
  2
- 0
+  nothing
  1
 
-julia> indexin(b,a)
-3-element Array{Int64,1}:
+julia> indexin(b, a)
+3-element Array{Union{Nothing, Int64},1}:
  6
  4
  3
 ```
 """
-function indexin(a::AbstractArray, b::AbstractArray)
-    bdict = Dict(zip(b, 1:length(b)))
-    [get(bdict, i, 0) for i in a]
+function indexin(a, b::AbstractArray)
+    bdict = Dict(zip(b, keys(b)))
+    map(i -> get(bdict, i, nothing), a)
```

Comment:
Until https://github.com/JuliaLang/julia/pull/25553 this is going to return an `Array{Any}`, but I guess it's OK as a temporary situation. We could specify the element type in advance, but then it would always be `Union{T, Nothing}` even in the absence of `nothing`.

Reply:
One benefit of `map` is that it returns a scalar if passed a scalar argument ([see here](https://github.com/JuliaLang/julia/pull/23845#issuecomment-332070524) for previous discussion).  So I think it makes sense to keep this as is and rely on #25553.

Also, I would have expected tests to fail due to the (currently) inaccurate types in the doctests, but it seems these doctests are not actually being tested.

> We could specify the element type in advance, but then it would always be Union{T, Nothing} even in the absence of nothing.

I (obviously) don't understand inference, but I would have expected/hoped that the return type would always be `Union{T, Nothing}`.  (In particular, I changed the type of the second doctest as well so this would be the case, despite lack of `nothing` in the output.)  If this is not expected, then the return type of this function will not be inferred, no?

Reply:
Inference isn't involved here, we just use `typejoin` to widen the element type as new elements are encountered. So no, the element type will always reflect what the array actually contains. This decision has been made so that inference can change without affecting user-visible behavior. We could force always returning a `Union{T, Nothing}` array, but I'm not sure that's a good idea given that `map` behaves differently.

Reply:
Seems like this may be an appropriate place to type-hint it as Union{valuetype(bdict), Nothing}

Reply:
You mean, just adding a type assertion? Wouldn't inference be able to figure that anyway?

Reply:
No, I mean with a typed-map:
```julia
function indexin(a, b::AbstractArray)
    indexes = keys(b)
    bdict = Dict(zip(b, indexes))
    return Union{eltype(indexes), Nothing}[
        get(bdict, i, nothing) for i in a
    ]
end
 ```

Reply:
OK. The question is then: why would be want to always use `Union{T, Nothing}` here, but have `map` change its return type depending on whether `nothing` actually appears or not?

Reply:
In brief, I think it all comes down to: `map` can't do better, so it should do the best it can with the information it has. This can do better, so it should.

Maybe it's just my personal preference and bias towards an "engineers null". I would summarize my current thinking on the Nullable-debate as: containers should propagate their element types explicitly, but value computations should not assert an output type (examples below).

Note however, that this doesn't apply to "user" code, which should "DWIM". I'm specifically talking about utility functions such as this one and most of the code in base and much of the code in packages – but not the application code and interactive usage they exist to support. (application example: use typed-map in reusable code with an obvious output type – like here – but provide easy access to the basic untyped-map)

Examples of the "container rule", currently existing:

`similar` - preserves eltype when changing container type, and is used as the basis of many of these other functions
`filter` - preserves the original array type
`collect` - e.g. `collect(values(::Dict{String, Union{Int, Nothing}))` returns a nullable-array
also c.f. `Base.IteratorEltype`
even the relatively complex looking `Dict(zip(b, keys(b)))` above can be seen to just preserve the eltype
counter-example: `Generator` can be good at finding iterators that aren't careful about this - for instance, in looking for examples, I noticed that iteration over IndexValue returns the incorrect type: the Pair type returned might not match its declared `eltype`; additionally iteration over `zip` returns a Tuple, so it is not always possible to precisely express the eltype of the result.

Examples of the "value rule", currently existing:

`tryparse` - returns T or Nothing
`get(d, x, nothing)` - returns T or Nothing
`findfirst` - returns eltype(keys(A)) or Nothing
counter-example: `Nullable`

But `map` is different from either of these. Since it does a computation, it can't simultaneously satisfy both of my "rules". Whereas typed-map / typed-comprehensions are fine, since the eltype to preserve is passed in an argument. So my belief is that the behavior of `map` is the precisely right tool for some tasks, and particularly ill-suited for others. In particular, if you're doing a computation, it is the simplest way to capture the output under the assumption that the computation has an (unknown) constant output type. If you're doing a copy or lookup (like here) and making a utility function (like here), then it just might not be the right tool for the job.

Reply:
Yeah, I agree that the present case is similar to copy/filter/lookup, so indeed it makes sense to preserve the original element type. But these issues are really tricky.

@garrison Could you change the PR to use Jameson's proposal then? Sorry for the back-and-forth.

Reply:
> typed-map / typed-comprehensions

I may be missing something, but is there a form of typed-map that is not a comprehension?  I'd like `indexin(4, [3,4,5])` to return `2`, not an `Array{Int64,0}` (as previously discussed [here](https://github.com/JuliaLang/julia/pull/23845#issuecomment-332070524)).

Reply:
(If there is no better solution, we could always go back to requiring `a::AbstractArray`.)

Reply:
I'm not sure why the result needs to be a scalar in that case. With a scalar argument, this becomes equivalent to `findlast` and building the Dict is not worthwhile. So that can be added as a special case, or just accept the 0-d array result.

Reply:
Yeah, probably not worrying about that special case.

Reply:
Though @StefanKarpinski [seemed to have a different opinion](https://github.com/JuliaLang/julia/pull/23845#issuecomment-332078223). Do you really want that behavior? At first returning a scalar made sense to me too, but thinking a bit more about the purpose of this function it's not so clear.

Anyway it would be easy to add a special method for `Number` if we want.

Reply:
Looking into this more, I'm honestly not sure why `Number` is treated differently than other data types for `map`, `collect`, etc.  First, there is (at least in my opinion) an impedance mismatch between `map` and `collect`:

```julia
julia> map(i->i, 3)
3

julia> collect(3)
0-dimensional Array{Int64,0}:
3
```

Additionally, I find this funny because either of these operations on a `Char` would result in a **1-d** Array.

Also pinging @mbauman, as he [had the opinion](https://github.com/JuliaLang/julia/pull/23845#issuecomment-332070524) as well that it should return a scalar.  As for me, I'm no longer convinced that it should (or is worth worrying about).

Reply:
Looks like `Char` is not `HasShape` even though `ndims` is defined for it.

Reply:
> As for me, I'm no longer convinced that it should (or is worth worrying about).

After re-reading that issue and this one, I think I would agree.  It stinks that we've never managed to remove iteration from numbers, but that's the real problem here.  We've papered over that in piecemeal cases as we've noticed it.  So my 👍 was just to continue that "papering over," but I'm becoming less convinced that's the direction we should go.  And of course, this ties into the big "what's a scalar" broadcast issue.

Reply:
Better discuss this in a separate issue? I'd rather do whatever sounds reasonable here and see how we can find a consistent general pattern later.